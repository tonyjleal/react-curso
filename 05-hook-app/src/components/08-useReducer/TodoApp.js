import React, { useReducer } from 'react';
import './style.css';
import { todoReducer } from './todoReducer';


export const TodoApp = () => {

 const init = () => {
  return [{
    id: new Date().getTime(),
    desc: 'Aprender React', 
    done: false
  }];
 }
 
 const [ todos, dispatch ]  = useReducer(todoReducer, [], init);


  const handleSubmit = (e) => {
    e.preventDefault();

    const newTodo = {
      id: new Date().getTime(),
      desc: 'New Task',
      done: false
    }
    
    const action = {
      type: 'add',
      payload: newTodo
    }

    dispatch(action);

  }


  return(
     <div>

        <h1> TODO APP ( { todos.length })</h1>
        <hr />
        <div  className='row'
        >
          <div className='col-7'>
            <ul className='list-group list-group-flush'>
              {
                todos.map((todo, i) => {
                  return <li
                          key={todo.id}
                          className='list-group-item'>
                          <p className='text-center'>{ i +1 }. { todo.desc }</p> 
                            <button className='btn btn-danger'>
                              Borrar
                            </button>
                          </li>
                })
              }
            </ul>
          </div>
          <div className='col-5'>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={handleSubmit}>
              <input 
              className='form-control'
              type="text" 
              name="description"
              placeholder='Aprender...'
              autoComplete='off' />

              <button type='submit' className='btn btn-outline-primary col-12 mt-1'>
                Agregar
              </button>
            </form>
          </div>

        </div>
    </div>
    );
};
