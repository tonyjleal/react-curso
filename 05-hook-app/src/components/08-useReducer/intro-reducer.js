
const initialState = [{
    id: 1,
    todo: 'Buy milk',
    done: false,
}]


const todoReducer = ( state = initialState, action ) => {

    if (action?.type === 'add') {
        return [...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Buy cookies',
    done: false,
}

const action = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, action)

console.log(todos);