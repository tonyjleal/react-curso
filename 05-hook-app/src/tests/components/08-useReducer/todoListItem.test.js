import {shallow} from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test <TodoListItem />', () => { 
    const todo = demoTodos[0];
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    
    const wrapper = shallow(
        <TodoListItem 
            key={todo.id}
            todo={todo}
            index={0}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
            />
    );

    test('should show component success', () => { 

        expect(wrapper).toMatchSnapshot();

    })
    
    test('should call function delete', () => { 
        
        wrapper.find('button').simulate('click')
        expect(handleDelete).toHaveBeenCalledWith( todo.id );
    })
    
    test('should call function toggle', () => { 
        
        wrapper.find('p').simulate('click')
        expect(handleToggle).toHaveBeenCalledWith( todo.id );
    })
    
    test('should show info success', () => { 

        const pClass = wrapper.find('p');

        expect(pClass.hasClass('false')).toBe(true);
        expect(pClass.text().trim()).toBe(`1. ${ todo.desc }`);

    })

    test('should have class complete if toggle is TODO.done is tue', () => { 


        todo.done = true;
        
        const wrapper = shallow(
            <TodoListItem 
                key={todo.id}
                todo={todo}
                index={0}
                handleDelete={handleDelete}
                handleToggle={handleToggle}
                />
        );
        const pClass = wrapper.find('p');

        expect(pClass.hasClass('complete')).toBe(true);
        expect(pClass.text().trim()).toBe(`1. ${ todo.desc }`);
    })
 })