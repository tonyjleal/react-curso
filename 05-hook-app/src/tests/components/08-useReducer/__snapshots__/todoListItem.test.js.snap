// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TodoListItem /> should show component success 1`] = `
<li
  className="list-group-item"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
